### PR TITLE
Custom location source fix

### PR DESCRIPTION
**Pull request by tobrun**
Follow-up https://github.com/mapbox/mapbox-gl-native/pull/9141. The CI bots didn't want to start for that external PR. 
Closes https://github.com/mapbox/mapbox-gl-native/issues/9139 and potentially https://github.com/mapbox/mapbox-gl-native/issues/8999

cc @LukasPaczos

### CODE REVIEW

File: platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/Mapbox.java [hunk 1]
```diff
@@ -55,9 +57,10 @@ public static synchronized Mapbox getInstance(@NonNull Context context, @NonNull
     return INSTANCE;
   }
 
-  Mapbox(@NonNull Context context, @NonNull String accessToken) {
+  Mapbox(@NonNull Context context, @NonNull String accessToken, LocationSource locationSource) {
     this.context = context;
     this.accessToken = accessToken;
+    this.lostLocationSource = locationSource;
```
Comment (zugaldia) [inline platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/Mapbox.java]: Any reason why the local `locationSource` is called differently (`lostLocationSource `)?

File: platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/constants/MapboxConstants.java [hunk 2]
```diff
@@ -133,5 +133,6 @@
   public static final String STATE_ATTRIBUTION_MARGIN_RIGHT = "mapbox_attrMarginRight";
   public static final String STATE_ATTRIBUTION_MARGIN_BOTTOM = "mapbox_atrrMarginBottom";
   public static final String STATE_ATTRIBUTION_ENABLED = "mapbox_atrrEnabled";
+  public static final String STATE_USING_CUSTOM_LOCATION_SOURCE = "mapbox_usingcCustomLocationSource";
```
Comment (zugaldia) [inline platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/constants/MapboxConstants.java]: Extra `c` between `mapbox_using` and `CustomLocationSource`.

File: platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/location/LocationSource.java [hunk 3]
```diff
@@ -68,8 +69,11 @@ public static synchronized LocationEngine getLocationEngine(@NonNull Context con
    */
   @Override
   public void activate() {
-    if (lostApiClient != null && !lostApiClient.isConnected()) {
+    Timber.e("LOCDEBUG %s activate - going to connect %s", hashCode(), !lostApiClient.isConnected());
```
Comment (zugaldia) [inline platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/location/LocationSource.java]: We shouldn't be logging an error every time the location source is activated.

File: platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/location/LocationSource.java [hunk 4]
```diff
@@ -68,8 +69,11 @@ public static synchronized LocationEngine getLocationEngine(@NonNull Context con
...
+    if (!lostApiClient.isConnected()) {
       lostApiClient.connect();
+    }else{
```
Comment (zugaldia) [inline platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/location/LocationSource.java]: nit: extra spacing.

File: platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/location/LocationSource.java [hunk 5]
```diff
@@ -68,8 +69,11 @@ public static synchronized LocationEngine getLocationEngine(@NonNull Context con
...
+      Timber.e("LOCDEBUG %s activate - NO connect", hashCode());
```
Comment (zugaldia) [inline platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/location/LocationSource.java]: Same comment as before, this probably should be removed.

File: platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/location/LocationSource.java [hunk 6]
```diff
@@ -80,7 +84,8 @@ public void activate() {
    */
   @Override
   public void deactivate() {
-    if (lostApiClient != null && lostApiClient.isConnected()) {
+    Timber.e("LOCDEBUG %s deactivate - going to disconnect %s", hashCode(), lostApiClient.isConnected());
```
Comment (zugaldia) [inline platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/location/LocationSource.java]: Same comment as before, this probably should be removed.

File: platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/location/LocationSource.java [hunk 7]
```diff
@@ -101,6 +106,7 @@ public boolean isConnected() {
    */
   @Override
   public void onConnected() {
+    Timber.e("LOCDEBUG %s onConnected", hashCode());
```
Comment (zugaldia) [inline platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/location/LocationSource.java]: Same comment as before, this probably should be removed.

File: platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/location/LocationSource.java [hunk 8]
```diff
@@ -112,6 +118,7 @@ public void onConnected() {
   @Override
   public void onConnectionSuspended() {
     Timber.d("Connection suspended.");
+    Timber.e("LOCDEBUG %s ConnectedSuspended", hashCode());
```
Comment (zugaldia) [inline platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/location/LocationSource.java]: Same comment as before, this probably should be removed.

File: platform/android/MapboxGLAndroidSDKTestApp/src/main/java/com/mapbox/mapboxsdk/testapp/activity/FeatureOverviewActivity.java [hunk 9]
```diff
@@ -119,36 +123,42 @@ private void startFeature(Feature feature) {
...
+      Snackbar.LENGTH_SHORT).show();
   }
 
   @Override
-  public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {
-    if (!isRuntimePermissionsRequired() || isPermissionAccepted(grantResults)) {
-      startFeature(features.get(requestCode));
-    } else {
+  public void onPermissionResult(boolean b) {
+    if(b){
```
Comment (zugaldia) [inline platform/android/MapboxGLAndroidSDKTestApp/src/main/java/com/mapbox/mapboxsdk/testapp/activity/FeatureOverviewActivity.java]: Let's give a proper name to this variable

File: platform/android/MapboxGLAndroidSDKTestApp/src/main/java/com/mapbox/mapboxsdk/testapp/activity/FeatureOverviewActivity.java [hunk 10]
```diff
@@ -119,36 +123,42 @@ private void startFeature(Feature feature) {
...
+      startFeature(features.get(locationActivityInList));
+    }else{
```
Comment (zugaldia) [inline platform/android/MapboxGLAndroidSDKTestApp/src/main/java/com/mapbox/mapboxsdk/testapp/activity/FeatureOverviewActivity.java]: nit: spacing

File: platform/android/MapboxGLAndroidSDKTestApp/src/main/java/com/mapbox/mapboxsdk/testapp/activity/FeatureOverviewActivity.java [hunk 11]
```diff
@@ -119,36 +123,42 @@ private void startFeature(Feature feature) {
...
-    return android.os.Build.VERSION.SDK_INT >= Build.VERSION_CODES.M;
+  @Override
+  public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {
+    super.onRequestPermissionsResult(requestCode, permissions, grantResults);
+    permissionsManager.onRequestPermissionsResult(requestCode, permissions, grantResults);
   }
 
-  private boolean isPermissionAccepted(@NonNull int[] grantResults) {
-    return grantResults.length > 0 && grantResults[0] == PackageManager.PERMISSION_GRANTED;
+  private boolean isRuntimePermissionsRequired() {
```
Comment (zugaldia) [inline platform/android/MapboxGLAndroidSDKTestApp/src/main/java/com/mapbox/mapboxsdk/testapp/activity/FeatureOverviewActivity.java]: How about we add this to the permissions manager? Let's ticket it for now.

File: platform/android/MapboxGLAndroidSDKTestApp/src/main/java/com/mapbox/mapboxsdk/testapp/activity/userlocation/BaseLocationActivity.java [hunk 12]
```diff
@@ -29,16 +39,21 @@ protected final void toggleGps(boolean enableGps) {
...
+  public void onExplanationNeeded(List<String> list) {
+    Snackbar.make(
+      findViewById(android.R.id.content),
+      TextUtils.join("",list.toArray()),
+      Snackbar.LENGTH_SHORT).show();
+  }
+
+  @Override
+  public void onPermissionResult(boolean b) {
+    enableLocation(b);
```
Comment (zugaldia) [inline platform/android/MapboxGLAndroidSDKTestApp/src/main/java/com/mapbox/mapboxsdk/testapp/activity/userlocation/BaseLocationActivity.java]: Same comment as before applies to all activities.

File: platform/android/MapboxGLAndroidSDKTestApp/src/main/java/com/mapbox/mapboxsdk/testapp/activity/userlocation/MockLocationEngine.java [hunk 13]
```diff
@@ -1,38 +1,52 @@
...
-
-  private Handler handler;
-  int currentIndex;
-
-  public MockLocationEngine() {
-    super();
+  public static synchronized MockLocationEngine getInstance() {
+    if (INSTANCE == null) {
+      INSTANCE = new MockLocationEngine(
+        MockLocationEngine.createLocation(40.416913, -3.703861),
```
Comment (zugaldia) [inline platform/android/MapboxGLAndroidSDKTestApp/src/main/java/com/mapbox/mapboxsdk/testapp/activity/userlocation/MockLocationEngine.java]: Any reason to use less locations than before?

File: platform/android/MapboxGLAndroidSDKTestApp/src/main/java/com/mapbox/mapboxsdk/testapp/activity/userlocation/MyLocationDrawableActivity.java [hunk 14]
```diff
@@ -66,29 +67,29 @@ public void onMapReady(MapboxMap map) {
...
-      if (location != null) {
-        onLocationChanged(location);
-      } else {
-        LocationSource.getLocationEngine(this).addLocationEngineListener(this);
-      }
-    } else {
-      mapboxMap.setMyLocationEnabled(false);
+    mapboxMap.setMyLocationEnabled(enabled);
+    if (lostApiClient == null) {
+      lostApiClient = new LostApiClient.Builder(this).addConnectionCallbacks(this).build();
```
Comment (zugaldia) [inline platform/android/MapboxGLAndroidSDKTestApp/src/main/java/com/mapbox/mapboxsdk/testapp/activity/userlocation/MyLocationDrawableActivity.java]: Why are we using the `LostApiClient` directly here? Would it be better to encapsulate under a custom `LocationEngine`?

File: platform/android/MapboxGLAndroidSDKTestApp/src/main/java/com/mapbox/mapboxsdk/testapp/activity/userlocation/MyLocationTintActivity.java [hunk 15]
```diff
@@ -133,7 +133,7 @@ public void onLocationChanged(Location location) {
   protected void onStart() {
     super.onStart();
     mapView.onStart();
-    LocationSource.getLocationEngine(this).addLocationEngineListener(this);
+//    LocationSource.getLocationEngine(this).addLocationEngineListener(this);
```
Comment (zugaldia) [inline platform/android/MapboxGLAndroidSDKTestApp/src/main/java/com/mapbox/mapboxsdk/testapp/activity/userlocation/MyLocationTintActivity.java]: Delete.

File: platform/android/MapboxGLAndroidSDKTestApp/src/main/java/com/mapbox/mapboxsdk/testapp/activity/userlocation/MyLocationTintActivity.java [hunk 16]
```diff
@@ -151,7 +151,7 @@ public void onPause() {
   @Override
   protected void onStop() {
     super.onStop();
-    LocationSource.getLocationEngine(this).removeLocationEngineListener(this);
+   // LocationSource.getLocationEngine(this).removeLocationEngineListener(this);
```
Comment (zugaldia) [inline platform/android/MapboxGLAndroidSDKTestApp/src/main/java/com/mapbox/mapboxsdk/testapp/activity/userlocation/MyLocationTintActivity.java]: Delete.

File: platform/android/MapboxGLAndroidSDKTestApp/src/main/java/com/mapbox/mapboxsdk/testapp/activity/userlocation/MyLocationTrackingModeActivity.java [hunk 17]
```diff
@@ -123,14 +128,24 @@ public void onMyBearingTrackingModeChange(@MyBearingTracking.Mode int myBearingT
           }
         });
 
+
```
Comment (zugaldia) [inline platform/android/MapboxGLAndroidSDKTestApp/src/main/java/com/mapbox/mapboxsdk/testapp/activity/userlocation/MyLocationTrackingModeActivity.java]: nit: extra line

File: platform/android/build.gradle [hunk 18]
```diff
@@ -1,9 +1,12 @@
 buildscript {
     repositories {
         jcenter()
+        maven {
+            url 'https://maven.google.com'
+        }
     }
     dependencies {
-        classpath 'com.android.tools.build:gradle:2.3.1'
+        classpath 'com.android.tools.build:gradle:3.0.0-alpha2'
```
Comment (zugaldia) [inline platform/android/build.gradle]: Let's avoid an alpha version here.

**Comment by zugaldia** [comment]
This PR is amazing, it touches and fixes so many places. Great job @LukasPaczos @tobrun.

**Comment by tobrun** [comment]
I have rebased the PR and working on integrating `LOST-v3.1.0-SNAPSHOT`

**Comment by tobrun** [comment]
a couple of items that remain to be tackled:
 - we removed LastKnownLocation from this PR as it would often result in a not very accurate location. Removal of this though makes the MyLocationView only show after the first location update. This also the case when toggling the MyLocationView on/off constantly. The UX around this should be improved. It now feels like the feature is broken and we aren't receiving any location updates.
 -  state is not correctly restored for: defining a custom LocationSource, replacing that source with the default one and do an orientation change. The custom location source is restored instead of the default one.

**Comment by LukasPaczos** [comment]
@tobrun Referring to former - if I'm not wrong it is enough to set any `Location` as last location to be returned by `getLastLocation()` on creation of custom `LocationEngine` or for example:
```
@Override
    public Location getLastLocation() {
        if (lastLocation == null)
            lastLocation = UserLocation.getLastKnownLocation();

        return lastLocation;
    }
```
I don't know how `LOST` (and therefore default engine) works in this regard.

And about latter - I'll look into this.

**Comment by tobrun** [comment]
@LukasPaczos thank you for :eyes: but already working on the the items above :wink:

**Comment by tobrun** [comment]
@LukasPaczos re. getLastKnownLocation -> this is what we initially had but weren't really happy with the results (the location was not very close to the actual position). I was expecting current UX to be better but it  isn't as it sometimes takes a while before we get a location fix. At this point I'm thinking of re hooking back into last known location but optimising that the following location update happens without any animation.

**Comment by LukasPaczos** [comment]
@tobrun What we have implemented is we are saving last location to `SharedPreferences` and when user restarts the app we are loading that, saved location and greying out `MyLocationView`, waiting for location fix.

**Comment by tobrun** [comment]
For privacy/security concerns, we don't write the user location to disk. For an app that approach can be justifiable but I feel it's out of scope for this library. On top of this I also think that it's the responsibility of the location provider to optimize this as much as possible.

**Comment by tobrun** [comment]
> state is not correctly restored for: defining a custom LocationSource, replacing that source with the default one and do an orientation change. The custom location source is restored instead of the default one.

Actually this is a working as designed in OnMapReady we are resetting it. End users could have more fine grain control over this by checking against savedInstanceState bundle. Calling this part resolved.

File: platform/android/MapboxGLAndroidSDKTestApp/src/main/java/com/mapbox/mapboxsdk/testapp/activity/FeatureOverviewActivity.java (hunk 11 above)
Comment (tobrun) [inline platform/android/MapboxGLAndroidSDKTestApp/src/main/java/com/mapbox/mapboxsdk/testapp/activity/FeatureOverviewActivity.java]: This has been ticketed out [here](https://github.com/mapbox/mapbox-java/issues/419#issuecomment-290442893), though the issue is still open.

File: platform/android/MapboxGLAndroidSDKTestApp/src/main/java/com/mapbox/mapboxsdk/testapp/activity/userlocation/MockLocationEngine.java (hunk 13 above)
Comment (tobrun) [inline platform/android/MapboxGLAndroidSDKTestApp/src/main/java/com/mapbox/mapboxsdk/testapp/activity/userlocation/MockLocationEngine.java]: We aren't using less, but actually using more. This MockLocationEngine is being powered by a ObjectAnimator that animates the location position over time. It produces 604 location updates to be precise.

File: platform/android/MapboxGLAndroidSDKTestApp/src/main/java/com/mapbox/mapboxsdk/testapp/activity/userlocation/MyLocationDrawableActivity.java (hunk 14 above)
Comment (tobrun) [inline platform/android/MapboxGLAndroidSDKTestApp/src/main/java/com/mapbox/mapboxsdk/testapp/activity/userlocation/MyLocationDrawableActivity.java]: This is to test what end users have been doing as well. That behaviour didn't show in the test app so including such an example hardens against regressions.

**Comment by tobrun** [comment]
Ready for another round of review @zugaldia

File: platform/android/MapboxGLAndroidSDKTestApp/src/main/java/com/mapbox/mapboxsdk/testapp/activity/userlocation/MockLocationEngine.java [hunk 19]
```diff
@@ -1,38 +1,55 @@
...
 import com.mapbox.services.android.telemetry.location.LocationEngine;
 import com.mapbox.services.android.telemetry.location.LocationEngineListener;
 
+import timber.log.Timber;
+
 /**
- * Sample LocationEngine that provides mocked locations simulating GPS updates
+ * Sample LocationEngine that provides mocked LOCATIONS simulating GPS updates
  */
 public class MockLocationEngine extends LocationEngine {
```
Comment (zugaldia) [inline platform/android/MapboxGLAndroidSDKTestApp/src/main/java/com/mapbox/mapboxsdk/testapp/activity/userlocation/MockLocationEngine.java]: @tobrun : @cammace is making the `MockLocationEngine` reusable across projects with https://github.com/mapbox/mapbox-java/pull/476. Could we use it here as well? If so, let's ticket a follow up ticket to make that happen.

File: platform/android/MapboxGLAndroidSDKTestApp/src/main/res/values/strings.xml [hunk 20]
```diff
@@ -151,6 +151,9 @@
     <string name="menuitem_title_reset">Reset</string>
     <string name="menuitem_title_rotate_gesture_enabled">Enable rotate gestures</string>
     <string name="menuitem_title_scroll_gesture_enabled">Enable scroll gestures</string>
+    <string name="menuitem_title_change_location_source_lost">Change to LOST location source</string>
+    <string name="menuitem_title_change_location_source_mock">Change to Mock location source</string>
```
Comment (zugaldia) [inline platform/android/MapboxGLAndroidSDKTestApp/src/main/res/values/strings.xml]: nit: `mock`

File: platform/android/dependencies.gradle [hunk 21]
```diff
@@ -21,8 +21,8 @@ ext {
             mapboxAndroidTelemetry : "com.mapbox.mapboxsdk:mapbox-android-telemetry:${mapboxServicesVersion}@aar",
 
             // mapzen lost
-            lost                   : 'com.mapzen.android:lost:3.0.0',
-
+            lost                   : 'com.mapzen.android:lost:3.0.1-SNAPSHOT',
```
Comment (zugaldia) [inline platform/android/dependencies.gradle]: @tobrun Is there any way to anchor this dependency to a specific SNAPSHOT timestamp? I'm afraid that depending on a moving target can make troubleshooting harder.

File: platform/android/MapboxGLAndroidSDKTestApp/src/main/java/com/mapbox/mapboxsdk/testapp/activity/userlocation/MockLocationEngine.java (hunk 19 above)
Comment (tobrun) [inline platform/android/MapboxGLAndroidSDKTestApp/src/main/java/com/mapbox/mapboxsdk/testapp/activity/userlocation/MockLocationEngine.java]: https://github.com/mapbox/mapbox-gl-native/issues/9226

File: platform/android/dependencies.gradle (hunk 21 above)
Comment (tobrun) [inline platform/android/dependencies.gradle]: It's a fairly new feature in gradle but it should be possible: https://docs.gradle.org/2.4/release-notes.html#depending-on-a-particular-maven-snapshot-version

**Pull request merged**